xde-tests/src/lib.rs

```diff
@@ -106,20 +244,75 @@ impl OptePort {
             gateway_mac: "a8:40:25:00:00:01".parse().unwrap(),
-            vni: Vni::new(1701u32).unwrap(),
+            vni: Vni::new(DEFAULT_MULTICAST_VNI).unwrap(),
+            phys_ip: phys_ip.parse().unwrap(),
+        };
+        let adm = OpteHdl::open()?;
+        adm.create_xde(name, cfg.clone(), DhcpCfg::default(), false)?;
+        Ok(OptePort {
+            name: name.into(),
+            cfg,
+            mcast_subscriptions: RefCell::new(Vec::new()),
+        })
+    }
+
+    /// Create a new OPTE port with dual-stack (IPv4 + IPv6) support.
+    pub fn new_dualstack(
+        name: &str,
+        private_ip_v4: &str,
+        private_ip_v6: &str,
+        guest_mac: &str,
+        phys_ip: &str,
+    ) -> Result<Self> {
+        let cfg = VpcCfg {
+            ip_cfg: IpCfg::DualStack {
+                ipv4: Ipv4Cfg {
+                    vpc_subnet: OVERLAY_NET.parse().unwrap(),
+                    private_ip: private_ip_v4.parse().unwrap(),
+                    gateway_ip: OVERLAY_GW.parse().unwrap(),
+                    external_ips: ExternalIpCfg {
+                        snat: Some(SNat4Cfg {
+                            external_ip: "1.2.3.4".parse().unwrap(),
+                            ports: 1000..=2000,
+                        }),
+                        ephemeral_ip: None,
+                        floating_ips: vec![],
+                    },
+                },
+                ipv6: Ipv6Cfg {
+                    vpc_subnet: OVERLAY_NET_V6.parse().unwrap(),
+                    private_ip: private_ip_v6.parse().unwrap(),
+                    gateway_ip: OVERLAY_GW_V6.parse().unwrap(),
+                    external_ips: ExternalIpCfg {
+                        snat: Some(SNat6Cfg {
+                            external_ip: "2001:db8::1".parse().unwrap(),
+                            ports: 4097..=8192,
+                        }),
+                        ephemeral_ip: None,
+                        floating_ips: vec![],
+                    },
+                },
+            },
+            guest_mac: guest_mac.parse().unwrap(),
+            gateway_mac: "a8:40:25:00:00:01".parse().unwrap(),
+            vni: Vni::new(DEFAULT_MULTICAST_VNI).unwrap(),
             phys_ip: phys_ip.parse().unwrap(),
         };
         let adm = OpteHdl::open()?;
         adm.create_xde(name, cfg.clone(), DhcpCfg::default(), false)?;
-        Ok(OptePort { name: name.into(), cfg })
+        Ok(OptePort {
+            name: name.into(),
+            cfg,
+            mcast_subscriptions: RefCell::new(Vec::new()),
+        })
     }
 
     /// Add an overlay routing entry to this port.
     pub fn add_router_entry(&self, dest: &str) -> Result<()> {
         let adm = OpteHdl::open()?;
         adm.add_router_entry(&AddRouterEntryReq {
             port_name: self.name.clone(),
-            dest: IpCidr::Ip4(format!("{}/32", dest).parse().unwrap()),
+            dest: IpCidr::Ip4(format!("{dest}/32").parse().unwrap()),
```

Comment:
Thanks for catching some of the `format!` inlining here -- Clippy must have been missed on this, my bad.